---
title: "Lab09:Structural Bioinformatics Pt.1 "
author: "Qihao Liu (U08901197)"
format: pdf
toc: TRUE
---

# Introduction to the RCSB Protein Data Bank (PDB)

## PDB statistics

The main database for structural biology is called the PDB. Let's have a look at what it contains:

First, import the CSV file from PDB
```{r}
stats <- read.csv("Data Export Summary.csv")

```
Notice that some of the supposedly numeric values are imputed as character in the original CSV file
So, if we try:
```{r, error=TRUE}
stats$X.ray
sum(stats$X.ray)
```
Notice that the numbers has "" around them because they are characters, and we receive a error when trying to calculate the sum of the X.ray structures

- Note: This is due to the "," in the numeric value

To fix this, we can try a different read function (need `install.packages("tidyverse")`)
```{r pdbstats, message=FALSE}
library("tidyverse")
stats <- read_csv("Data Export Summary.csv")
```
Now we try sum the structures solved by x.ray again, and it works:

```{r}
sum(stats$`X-ray`)
```


>Q1: What percentage of structures in the PDB are solved by X-Ray and Electron Microscopy.

>A1: Percent of structures solved by x-ray is 81.48%, percent of structures solved by EM is 12.22%

```{r PercetXrayEM}
#Structures solved by x-ray
x_ray <- (sum(stats$`X-ray`)/sum(stats$`Total`))*100
round(x_ray,2)
#Structures solved by EM
em <- (sum(stats$`EM`)/sum(stats$`Total`))*100
round(em,2)
```

>Q2: What proportion of structures in the PDB are protein?

```{r PerctProtein, message=FALSE}
protein <- round((sum(stats[1,2:8])/sum(stats$`Total`))*100,2)
```

>A2: `r protein`% of structures are proteins


>Q3: Type HIV in the PDB website search box on the home page and determine how many HIV-1 protease structures are in the current PDB?

>A3: This query matches 4,865 Structures.

## The PDB format

No let's try to examine some structures from PDB. Starting off by downloading the 1HSG (HIV Protease) from PDB. This is the first protein of HIV targeted by drug.

First we load a package for structural bioinformatics, bio3d. The read.pdb function in this package allows us to get the PDB file for a structure from the PDB website.
```{r, message=FALSE}
library("bio3d")
hiv <- read.pdb("1HSG")
hiv
```

## Visualizing the HIV-1 protease structure

### Using Mol*

Let's first use the Mol* viewer to explore this structure (Mol* is a 3D structure viewer runs online)

- Go to https://molstar.org/, input the PDB code, and apply
- Notice the MK1 compound in the center of the HIV protease
- Download a preliminary image of the structure, save the png. file in the project directory, include it in the lab sheets
![](1HSG.png)

### Getting to know HIV-Pr

There is a water molecue and an aspartate residue (Asp25, D25) that is important for the catalytic activity of HIV-Pr. Find them in the structure

![](1HSG_H2O_D25.png)
>Q4: Water molecules normally have 3 atoms. Why do we see just one atom per water molecule in this structure?

>A4: Hydrogen is omitted from the H2O molecule because it is often not resolved

>Q5: There is a critical “conserved” water molecule in the binding site. Can you identify this water molecule? What residue number does this water molecule have

>A5:311

>Q6: Generate and save a figure clearly showing the two distinct chains of HIV-protease along with the ligand. You might also consider showing the catalytic residues ASP 25 in each chain and the critical water (we recommend “Ball & Stick” for these side-chains). Add this figure to your Quarto document.

![](1HSG_H2O_D25.png)

>Discussion Topic: Can you think of a way in which indinavir, or even larger ligands and substrates, could enter the binding site?

>A: Because HIV-pr is formed by two homodimers, larger ligands can either enter the catalytic site when the two monomers forms a homodimer or it can also enter the binding site when the two monomers changes comformation in their active state, which provides passage into the active site.

>Q7: [Optional] As you have hopefully observed HIV protease is a homodimer (i.e. it is composed of two identical chains). With the aid of the graphic display can you identify secondary structure elements that are likely to only form in the dimer rather than the monomer?

>A7: The parallel betasheets on the C-terminus formed by two monomers is a structure elments that are likely to only form in the dimer rather than the monomer. Same goes to the catalytic site containing two aspartate residue, which only forms when the homodimer comes together.

# Introduction to Bio3D in R

##Reading PDB file data into R

```{r}
library("bio3d")
hiv <- read.pdb("1HSG")
head(hiv$atom)
```

>Q7: How many amino acid residues are there in this pdb object?

>A7: There are a total of 198 protein residues in this pdb project

>Q8: Name one of the two non-protein residue?

>A8: HOH

>Q9: How many protein chains are in the structure

>A9: 2

Extract the sequence
```{r}
pdbseq(hiv)
```
Note that there are two chains included in this printed sequence.
So we are gonna get only sequence from chain A
```{r}
trim.pdb(hiv,chain="A")
chainA_seq <- pdbseq(trim.pdb(hiv,chain="A"))
```

## Quick PDB visualization in R
Next, install the **bio3dview** packages on CRAN so we can interactively view these PDB objets in R

```{r}
library(bio3dview)
library(NGLVieweR)
```


```{r error = TRUE, eval = FALSE, echo = TRUE}

view.pdb(hiv)

view.pdb(hiv, colorScheme = "sse", backgroundColor = "pink")

#Selecting ASP25 Residue
sel <- atom.select(hiv,resno=25)
view.pdb(hiv, highlight = sel,
         colorScheme = "sse",
        backgroundColor = "pink")
```

## Predict Functional Motion

```{r message=FALSE}
library(bio3d)
#Read Structure
adk <- read.pdb("6s36")
adk
#Run calculation
m <- nma(adk)
m
plot(m)
```
Generate a "trajectory" of predicted motion
```{r error = TRUE, eval = FALSE, echo = TRUE}
mktrj(m,file="ADK_nma.pdb")
view.nma(m, pdb=adk)
```

# Comparative Structure Analysis of Adenylate Kinase
The goal of this section is to perform principal component analysis (PCA) on the complete collection of Adenylate kinase structures in the protein data-bank (PDB).

## Set up
>Q10. Which of the packages above is found only on BioConductor and not CRAN?

>A10: msa

>Q11. Which of the above packages is not found on BioConductor or CRAN?: 

>A11: bio3dview

>Q12. True or False? Functions from the pak package can be used to install packages from GitHub and BitBucket? 

>A12: TRUE

## Search and retrieve ADK structures
Get ADK sequence from PDB

```{r}
library(bio3d)
aa <- get.seq("1ake_A")
aa
```
> Q13. How many amino acids are in this sequence, i.e. how long is this sequence? 

> A13:214

Now we can use this sequence as a query to BLAST search the PDB to find similar sequences and structures.
```{r}
# Blast or hmmer search 
#b <- blast.pdb(aa)
#hits <- plot(b)
#head(hits$pdb.id)
```
Unable to access BLAST service at the time, try alternative code:

```{r}
hits <- NULL
hits$pdb.id <- c('1AKE_A','6S36_A','6RZE_A','3HPR_A','1E4V_A','5EJE_A','1E4Y_A','3X2S_A','6HAP_A','6HAM_A','4K46_A','3GMT_A','4PZL_A')
#Download related PDB files
files <- get.pdb(hits$pdb.id, path="pdbs", split=TRUE, gzip=TRUE)
```
## Align and superpose structures

Next, use `pdbaln()` to align and superpose the identified PDB structures

```{r, fig.width=8, fig.height=6, fig.retina=2}
# Align releated PDBs
pdbs <- pdbaln(files, fit = TRUE, exefile="msa")

# Vector containing PDB codes for figure axis
ids <- basename.pdb(pdbs$id)

# Draw schematic alignment
plot(pdbs, labels=ids)
```

### Optional: Viewing our superposed structures

```{r error = TRUE, eval = FALSE, echo = TRUE}
library(bio3dview)

view.pdbs(pdbs)
```

## Annotate collected PDB structures
`pdb.annotate()` allow us to annotate the PDB files we have downloaded

```{r}
anno <- pdb.annotate(ids)
unique(anno$source)
head(anno)
```

## PCA

PCA can be performed on the structural ensemble (stored in the pdbs object)

```{r}
# Perform PCA
pc.xray <- pca(pdbs)
pc.xray
plot(pc.xray)
```

Function rmsd() will calculate all pairwise RMSD values of the structural ensemble. This facilitates clustering analysis based on the pairwise structural deviation:

```{r}
# Calculate RMSD
rd <- rmsd(pdbs)

# Structure-based clustering
hc.rd <- hclust(dist(rd))
grps.rd <- cutree(hc.rd, k=3)

plot(pc.xray, 1:2, col="grey50", bg=grps.rd, pch=21, cex=1)
```
This plot is obtained by projecting the individual structures onto two selected PCs (e.g. PC-1 and PC-2). These projections display the inter-conformer relationship in terms of the conformational differences described by the selected PCs.

# Optional further visualization

To visualize the major structural variations in the ensemble the function mktrj() can be used to generate a trajectory PDB file

```{r}
# Visualize first principal component
pc1 <- mktrj(pc.xray, pc=1, file="pc_1.pdb")
```

```{r error = TRUE, eval = FALSE, echo = TRUE}
view.pca(pc.xray)
```


We can also plot our main PCA results with ggplot:
```{r}
#Plotting results with ggplot2
library(ggplot2)
library(ggrepel)

df <- data.frame(PC1=pc.xray$z[,1], 
                 PC2=pc.xray$z[,2], 
                 col=as.factor(grps.rd),
                 ids=ids)

p <- ggplot(df) + 
  aes(PC1, PC2, col=col, label=ids) +
  geom_point(size=2) +
  geom_text_repel(max.overlaps = 20) +
  theme(legend.position = "none")
p
```

# Normal mode analysis

```{r}
# NMA of all structures
modes <- nma(pdbs)
plot(modes, pdbs, col=grps.rd)
```

>Q14. What do you note about this plot? Are the black and colored lines similar or different? Where do you think they differ most and why?

>A14: There are two distinct groups of structures that can be identified from this plot, which corresponds pretty well to our clusters identified from PCA. The black and colored lines are very different, they corresponds to the active state of ADK and inactive state of ADK(due to inhibitor, catalytically dead mutation etc.)
